### PR TITLE
Fix slice viewer crash

### LINKED ISSUE #15161: SliceViewer Beta Test Issue
SliceViewer reports std::bad_alloc dialog in rebin mode.

The basic problem here is that the system doesn't have enough memory, but, we should prevent the creation of large datasets if we know it's going to lead to out of memory issues.

Pascal demonstrated the issue to me on windows. I think you should be able to reproduce the issue as follows. 

**To Reproduce**
Probably most useful to be running in a Debug build on Windows

1. Open the task manager.
1. In MantidPlot run the SXD_NaCl.py example
1. Check the amount of free memory
1. Run `CloneMDWorkspace` until the amount of memory free is down to a few GB.
1. Open on of the MDEventWorkspaces (QLab) in the SliceViewer
1. Enter the Rebin mode. Adjust the #bins properties to yeild a larger workspace.
1. Try moving the sliders until you can reproduce the exception.

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
Fixes #15161 

**Issue:**
The slice viewer crashed when using rebin mode with very fine rebinning. The memory maxes out and the underlying vectors of the rebinned workspace cannot be allocated. This leads to an inconsistent rebin state in the slice viewer. 

When the user presses the zoom button, the color range and subsequently the data range are updated in the colorScaleAutoSlice method(https://github.com/mantidproject/mantid/blob/master/MantidQt/SliceViewer/src/SliceViewer.cpp#L849:L853). The issue starts in the `findRangeSlice` method as we are in a state where the rebinMode is `true` yet the overlay workspace `m_overlayWS` is `nullptr`(rebinning failed, hence we never got a rebinned workspace). The current logic does not evaluate the color range in this case. This means that the last value of `m_colorRangeSlice` will be used. If this the first time that this variable is being used, then we end up with the default setting of this variable. It turns out that the default is a `QwtDoubleInterval` with min= 0 and max = -1 which is an invalid interval range. Further down the line this color scale range is written into the data range of the `QwtRasterDataMD` data set which is used for plotting. This eventually will cause problems in Qt and Qwt (I did not assess this during the debug session though).

**Solution:**
In the `findRangeSlice` method we check if the Slive Viewer is in an inconsitent state, ie if `m_overlay` == nullptr and the rebin mode is set to true (or `m_overlay` exists but the rebin mode is false). If the state is consistent, the logic does not change. If the state is inconsistent, then we turn off the rebin mode and use the underlying workspace to extract the color scale range.
# For Testing
1. Load the NaCl event mode data set
2. Open the workspace in the Slice Viewer ( make sure that you dont press the zoom button)
3. Select the rebin mode 
4. Set rebinning to 9500 for both axes
   - Note that there will be warnings regarding `execution terminated`: This is ok
   - Increase the rebinning until you see an error saying `Bad Alloc`: This is to be expected
5. Once you see the `Bad Alloc` error in the Mantid console, click on the zoom button
   - Confirm that there is no exception

Please perform more free testing.

### CODE REVIEW

**Comment by picatess** [comment]
@VickieLynch  This PR solves a Slice Viewer crash. Unfortunately I am not very familiar with the Slice Viewer, especially with the different use cases. Owen mentioned that you have been using the Slice Viewer quite heavily in the  past and that you might be able to detect potential side effects from this fix. If you find the time it would be very helpful if you could have a play with this version. Thanks.

File: MantidQt/SliceViewer/src/SliceViewer.cpp [hunk 1]
```diff
@@ -65,6 +65,18 @@ using Poco::XML::NodeIterator;
 using Poco::XML::NodeFilter;
 using MantidQt::API::AlgorithmRunner;
 
+namespace {
+/*
+Checks if the rebinning is in a consistent state, ie if rebin mode is selected and there
+is a rebin workspace or there is no rebin workspace and no rebin mode selected.
+The state is inconsistent if rebin mode is selected and there is no workspace.
+*/
+  bool isRebinInConsistentState(Mantid::API::IMDWorkspace_sptr rebinnedWS, bool useRebinMode) {
```
Comment (OwenArnold) [inline MantidQt/SliceViewer/src/SliceViewer.cpp]: My only comment here (not worth changing) is that passing the shared_ptr implies additional, or change of ownership and that's not what's happening here. Would be better to pass `Mantid::API::IMDWorkspace_sptr&`. Furthermore, use the `.get` accessor and compare to null_ptr might be have made this more explicit as to what's being tested.

File: MantidQt/SliceViewer/src/SliceViewer.cpp [hunk 2]
```diff
@@ -1378,7 +1390,13 @@ part of the workspace */
 void SliceViewer::findRangeSlice() {
   IMDWorkspace_sptr workspace_used = m_ws;
   if (m_rebinMode) {
-    workspace_used = this->m_overlayWS;
+    // If the rebinned state is inconsistent, then we turn off
+    // the rebin selection and continue to use the original WS
+    if (!isRebinInConsistentState(m_overlayWS, m_rebinMode)) {
```
Comment (OwenArnold) [inline MantidQt/SliceViewer/src/SliceViewer.cpp]: This looks good and readable.

**Comment by OwenArnold** [comment]
@picatess . Could you please update the release notes at the end of this.

**Comment by OwenArnold** [comment]
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000257632000

Thread 26 Crashed:
0   libMantidDataObjects.dylib    	0x000000011900ec40 Mantid::DataObjects::MDHistoWorkspace::setTo(double, double, double) + 128
1   libMantidMDAlgorithms.dylib   	0x00000001250c6875 void Mantid::MDAlgorithms::BinMD::binByIterating<Mantid::DataObjects::MDEvent<3ul>, 3ul>(Mantid::DataObjects::MDEventWorkspace<Mantid::DataObjects::MDEvent<3ul>, 3ul>::sptr) + 261
2   libMantidMDAlgorithms.dylib   	0x00000001250c218e Mantid::MDAlgorithms::BinMD::exec() + 2222
3   libMantidAPI.dylib            	0x000000010f5da399 Mantid::API::Algorithm::execute() + 1721
4   libMantidAPI.dylib            	0x000000010f5e207c Mantid::API::Algorithm::executeAsyncImpl(Poco::Void const&) + 28
5   libMantidAPI.dylib            	0x000000010f5fc183 Mantid::API::AlgorithmProxy::executeAsyncImpl(Poco::Void const&) + 35
6   libMantidAPI.dylib            	0x000000010f5fdde0 Poco::ActiveRunnable<bool, Poco::Void, Mantid::API::AlgorithmProxy>::run() + 64
7   libPocoFoundation.dylib       	0x000000011001cf56 Poco::PooledThread::run() + 116
8   libPocoFoundation.dylib       	0x000000011001aca0 Poco::ThreadImpl::runnableEntry(void*) + 98
9   libsystem_pthread.dylib       	0x00007fff8f535268 _pthread_body + 131
10  libsystem_pthread.dylib       	0x00007fff8f5351e5 _pthread_start + 176
11  libsystem_pthread.dylib       	0x00007fff8f53341d thread_start + 13

**Comment by picatess** [comment]
* The threading issue could not be reproduced on Windows and Ubuntu)
* Owen's suggestions were included
* The `AutoScale` label was changed to `AutoScale on Load`
* The color scale range is set to the full range when loading a workspace (irrespective of the status of the `AutoScale on Load` checkbox). This restores the old behaviour and seems sensible even for Live Data

**Comment by OwenArnold** [comment]
@picatess Yes, I created this for the threading issue. https://github.com/mantidproject/mantid/issues/15199

File: MantidQt/SliceViewer/src/SliceViewer.cpp [hunk 3]
```diff
@@ -743,7 +755,7 @@ void SliceViewer::setWorkspace(Mantid::API::IMDWorkspace_sptr ws) {
   this->updateDimensionSliceWidgets();
 
   // Only autoscale color bar if box is checked
```
Comment (OwenArnold) [inline MantidQt/SliceViewer/src/SliceViewer.cpp]: Would be nice to update this comment at some point.

File: MantidQt/SliceViewer/src/SliceViewer.cpp [hunk 4]
```diff
@@ -743,7 +755,7 @@ void SliceViewer::setWorkspace(Mantid::API::IMDWorkspace_sptr ws) {
...
-  if (m_colorBar->getAutoScale()) {
+  if (!m_firstWorkspaceOpen || m_colorBar->getAutoScale()) {
```
Comment (OwenArnold) [inline MantidQt/SliceViewer/src/SliceViewer.cpp]: Can we re-write this slightly to make it easier to read. How about a non-member function something like `bool shouldAutoScale(isFirstWorkspaceOpen, isAutoscalingOnLoad)` ?

**Comment by OwenArnold** [comment]
Functionally I think this is right. Once Anton has made some minor changes could @VickieLynch also double check this?

**Comment by VickieLynch** [comment]
I could not get it to fail with SNS data rebinning large event data with fine grid.  Also the autoscale is working well again.

**Pull request merged**